hw/ip/otbn/dv/rig/rig/gens/misaligned_load_store.py

```diff
@@ -0,0 +1,290 @@
+
+        # Initialize the address value for the while loop below
+        addr = imm_val + value
+
+        # Randomize immediate until getting a misaligned address which is in a
+        # valid range otherwise
+        while ((not addr % 4) or addr < 0 or model.dmem_size <= addr):
+            imm_val = random.choice(range(min_offset, max_offset))
+            addr = imm_val + value
+
+        offset_val = imm_op_type.op_val_to_enc_val(imm_val, model.pc)
+        assert offset_val is not None
+
+        if insn.mnemonic == 'lw':
+            # Pick grd randomly. Since we can write to any register (including x0)
+            # this should always succeed.
+            op_val_grd = model.pick_operand_value(grd_op_type)
+            assert op_val_grd is not None
+
+            op_val = [op_val_grd, offset_val, op_val_grs1]
+
+        elif insn.mnemonic == 'sw':
+            # Any known register is okay for grs2 operand since it will be
+            # guaranteed to have a misaligned address to store the value from.
+            # We definitely have a known register (x0), so this should never fail
+            op_val_grs2 = random.choice(known_regs)[0]
+            assert op_val_grs2 is not None
+
+            op_val = [op_val_grs2, offset_val, op_val_grs1]
+
+        return ProgInsn(insn, op_val, ('dmem', 4096))
+
+    def _fill_bn_xid(self, insn: Insn, model: Model) -> Optional[ProgInsn]:
+        '''Fill out a BN.LID or BN.SID instruction'''
+
+        bn_imm_op_type = self.bnlid.operands[2].op_type
+
+        # Determine the offset range for BN.XID
+        bn_offset_rng = bn_imm_op_type.get_op_val_range(model.pc)
+        assert bn_offset_rng is not None
+
+        # Max/Min Offsets for BN.XID (-512 * 32, 511 * 32)
+        min_offset, max_offset = bn_offset_rng
+
+        # Get known registers
+        known_regs = model.regs_with_known_vals('gpr')
+
+        base = []
+
+        # Get misaligned registers which can easily be in valid address range
+        for reg_idx, reg_val in known_regs:
+            val = reg_val - (1 << 32) if reg_val << 31 else reg_val
+            if (((1 - max_offset <= val < model.dmem_size - min_offset) and
+                 val % 32)):
+                base.append((reg_idx, val))
+
+        if not base:
+            return None
+
+        # Pick a random register among known registers for initialization
+        idx, value = random.choice(base)
+
+        # Pick a random immediate and calculate address for initialization
+        bn_imm_val = random.randrange(min_offset, max_offset, 32)
+        addr = bn_imm_val + value
+
+        # Randomize both immediate and value until getting a misaligned address
+        # which is in a valid range otherwise
+        while ((not addr % 32) or 0 > addr or addr >= model.dmem_size):
+            bn_imm_val = random.randrange(min_offset, max_offset, 32)
+            idx, value = random.choice(base)
+            addr = bn_imm_val + value
```

Comment:
I think you can do this without the loop, and it might be a bit easier to understand what's going on? (untested, so sorry if this does something stupid!)
```suggestion
        # Pick a random representable immediate which yields an address in
        # range.
        min_imm = max(min_offset, 1 - value)
        max_imm = min(max_offset, model.dmem_size - 1 - value)

        imm_shift = bn_imm_op_type.shift
        min_imm_sh = (min_imm + (1 << imm_shift) - 1) >> imm_shift
        max_imm_sh = max_imm >> imm_shift
        assert min_imm_sh <= max_imm_sh

        bn_imm_val = random.randint(min_imm_sh, max_imm_sh) << imm_shift
        addr = bn_imm_val + value
        assert addr % 32
        assert 0 <= addr < model.dmem_size

        bn_offset_val = bn_imm_op_type.op_val_to_enc_val(bn_imm_val, model.pc)
        assert bn_offset_val is not None
```